### ISSUE TITLE
Recorded history and undo/redo not working properly

### ISSUE THREAD

**Issue by gngugnniokl** [opentoonz/opentoonz#4964]
### Description

After a while of using opentoonz thespecially when using mypaint brushes. Opentoonz stops recording history and undo stops undoing the most recent stroke

### Steps to Reproduce

1. Create multiple layers all using raster brushes.
2. draw brush strokes for at least 30 frames
3. open the history window
4. See if opentoonz still records history
5. try to undo

### Expected Behavior

Opentoonz will stop recording most recent brush strokes and undo the storke or action beforehand

### OpenToonz Version

Nightly (Specify below)

### Operating System

Windows

### GPU

Intel UHD integrated graphics

### Graphics Tablet

Huion

**Comment by RattusaurusRex** [comment]
I can't replicate this bug at all by following your instructions.  Is this bug easy for you to reproduce?  If so, can you try to record your screen while you demonstrate the bug?  Are you copying and pasting at all, because an undo bug was recently fixed that involved copying and pasting selections on Raster levels?

**Comment by gngugnniokl** [comment]
No. Im not copy pasting. I'll try to recreate it.

**Comment by gngugnniokl** [comment]
https://github.com/opentoonz/opentoonz/assets/75842539/08496174-3fc4-42c2-9485-79fdf419d549

**Comment by RattusaurusRex** [comment]
Looks to me like that was triggered by the auto-save with an untitled scene.  To avoid this for now, I would recommend either naming and saving the scene before creating and drawing on levels, or turning off auto-save.

**Comment by RodneyBaker** [comment]
@RattusaurusRex might be on to something here.
If you have autosave on then you probably will want to turn the option to clear Undo History when levels are saved to 'off':

![image](https://github.com/opentoonz/opentoonz/assets/2576432/a10b99b4-9f6f-4e65-a6e4-540fea4378aa)

**Comment by RattusaurusRex** [comment]
When testing, I turned the auto-save on 1 minute.  With an untitled scene, and clicking on the X for the save scene pop-up like you do in your video, the undo history got messed up every single time.   It doesn't matter how many layers or frames you have.  If the scene was saved first, there were no issues with the undo history.  If auto-save is off, there are also no issues with the undo history.

**Comment by gngugnniokl** [comment]
Hm. I tried to replicate what you did however it didn't seem to be triggered by autosave at all but rather strangely it seemed to be triggered after I used the keyframe tool.

**Comment by gngugnniokl** [comment]
Or even any effects at all for that matter. Undo history sadly seems to not work at all when working with complex effects and tweens. It renders the program unuseable unfortunately :-(

**Comment by RodneyBaker** [comment]
Undo/Redo and its history is an ongoing area of interest and development.
Some improvements have been made and others tend to be created as code is updated.

Perhaps a good case in point for this is with Tahoma2D where new features and enhancements touch upon older code.
There @manongjohn tends to add in the required code the processes require to successfully add to history and thereby give the program something appropriate to undo/redo.

Will recommend continuing to document shortfalls in history/undo/redo so they can be addressed as code is updated.

**Comment by RodneyBaker** [comment]
Closing this as needing to be updated which will help us resolve the broader scope of issues with Undo / History.  While we may not get that broader problem solved by version 1.8 we should be able to make good progress moving forward.

TLDR:  There are many issues with Undo / History and we need a deep dive to resolve those.
There area also a lot of moving parts.
